documentation/projects/proposals/ingestion_server_removal/20240319-project_proposal.md

```diff
@@ -0,0 +1,105 @@
+# 2024-03-19 Project Proposal: Removal of the Ingestion Server
+
+**Author**: @stacimc
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @Aetherunbound
+- [ ] @zackkrida
+
+## Project summary
+
+<!-- A brief one or two sentence summary of the project's features -->
+
+Currently, our data refresh process is _orchestrated_ by Airflow, but the work
+is actually performed on a bespoke ingestion server. This server is relatively
+complex compared to our needs, difficult to maintain, and makes the data refresh
+process unnecessarily opaque. Because the refresh is triggered in a single task
+on the ingestion server, it is not possible to track the progress in Airflow or
+retry it from a point of failure. Progress must be followed through periodic
+Slack notifications and inspection of logs in Cloudwatch.
+
+In this project, we will move all operational pieces of the data refresh into
+Airflow and completely remove the ingestion server.
+
+## Goals
+
+<!-- Which yearly goal does this project advance? -->
+
+Moving the process entirely into Airflow will have the following advantages for
+the critical data refresh process:
+
+- Easier iteration, as changes to the implementation can be deployed immediately
+  by simply merging code in the catalog
+- Increased reusability of code, particularly among the growing number of
+  Elasticsearch DAGs
+- Increased observability of the process: we'll have the ability to see exactly
+  what step the refresh is on by inspecting the running DAG, with much more
+  granularity even than the existing Slack notifications. This also includes
+  much more convenient access to timing information
+- Ability to pause the process before the next step starts, or retry a failed
```

Comment:
This sounds incredible! Do you think this would require additional work? To clarify what I mean, would this be a "free" upgrade as a consequence of porting over the ingestion server, or something we would intentionally need to devote development time to achieving?

Reply:
Sure! I'll add more context to the doc, but yes this is something we would get for free. The only caveat is related to pausing -- we can pause the DAG but it will typically not pause whatever step we're currently on, if it's something that has been triggered on a remote service (like the copy data query running on the API DB or an indexer worker chugging through indexing). However it _will_ prevent the _next_ steps from running, so for example we could pause the DAG before the reindexing steps start to prevent them from happening.

Retrying is even more exciting because we should be able to do things like retry failed reindexing steps without having to start over from the beginning, or even retry an individual failed indexer worker directly from Airflow.